packages/driver/src/util/serialization/log.ts

```diff
@@ -0,0 +1,391 @@
+ *
+ * @param logAttrs raw log attributes passed in from either a log:changed or log:added event
+ * @returns a serializable form of the log, including attempted serialization of DOM elements and Functions (if explicitly stated)
+ */
+export const preprocessLogForSerialization = (logAttrs) => {
+  let { snapshots, ...logAttrsRest } = logAttrs
+
+  const preprocessed = preprocessLogLikeForSerialization(logAttrsRest)
+
+  if (preprocessed) {
+    if (snapshots) {
+      preprocessed.snapshots = snapshots.map((snapshot) => preprocessSnapshotForSerialization(snapshot))
+    }
+
+    if (logAttrs?.consoleProps?.table) {
+      preprocessed.consoleProps.table = preprocessLogLikeForSerialization(logAttrs.consoleProps.table, true)
+    }
+  }
+
+  return preprocessed
+}
+
+/**
+ * Re defines log messages being received in the primary domain before sending them out through the event-manager.
+ *
+ * Efforts here include importing captured snapshots from the spec bridge into the primary snapshot document, importing inline
+ * snapshot styles into the snapshot css map, and reconstructing DOM elements and functions somewhat naively.
+ *
+ * To property render consoleProps/$el elements in snapshots or the console, DOM elements are lazily calculated via
+ * getter properties on an object. If these DOM elements are in an array, the array is defined as an ES6 proxy that
+ * ultimately proxies to these getter objects.
+ *
+ * The secret here is that consoleProp DOM elements needs to be reified at console printing runtime AFTER the serialized snapshot
+ * is attached to the DOM so the element can be located and displayed properly
+ *
+ * In most cases, the element can be queried by attributes that exist specifically on the element or by the `HIGHLIGHT_ATTR`. If that fails or does not locate an element,
+ * then a new element is created against the snapshot context. This element will NOT be found on the page, but will represent what the element
+ * looked like at the time of the snapshot.
+ *
+ * @param logAttrs serialized/preprocessed log attributes passed to the primary domain from a spec bridge
+ * @returns a reified version of what a log is supposed to look like in Cypress
+ */
+export const reifyLogFromSerialization = (logAttrs) => {
+  let { snapshots, ... logAttrsRest } = logAttrs
+
+  if (snapshots) {
+    // @ts-ignore
+    snapshots = snapshots.filter((snapshot) => !!snapshot).map((snapshot) => reifySnapshotFromSerialization(snapshot))
+  }
+
+  const reified = reifyLogLikeFromSerialization(logAttrsRest)
+
+  if (reified.$el && reified.$el.length) {
+    // make sure $els are jQuery Arrays to keep was is expected in the log
+    reified.$el = Cypress.$(reified.$el.map((el) => el))
+  }
+
+  reified.snapshots = snapshots
+
+  // use this flag to help reify the log correctly when certain property types slightly differ
+  reified.crossOriginLog = true
```

Comment:
right now we are only using this flag to properly store the `$el` in the log after reification.